keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,781 @@
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine learning user, ML frameworks allow scheduling of multiple pods.  
+The Job controller does not typically wait for terminating pods to be marked as failed.  Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.  The following yaml can fit these needs:
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+#### Story 3
+
+As a cloud user, users would want to guarantee that the number of pods that are running includes terminating pods.  In scare compute environments, users may only have a limited amount of nodes and they do not want to try and schedule pods to a new resource.
+Counting terminating pods as active allows for the scheduling of pods to wait until pods are terminated.
+
+See [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
+
+Another open question is if we want to include Deployments in the initial release of this feature.  There is some discussion about releasing the Job API first and then follow up with Deployment.  
+
+We decided to define the APIs in this KEP as they can utilize the same implementation.  
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new BoolPtr field:
+
+```golang
+type JobSpec struct{
+  ...
+ // terminatingAsActive specifies if the Job controller should include terminating pods
+ // as active. If the field is true, then the Job controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
```

Comment:
Not in love with this name, is there a list of alternative names or a link to a discussion of naming?

Reply:
I might prefer `activeUntilTerminal` because the “active” prefix sorts it with activeDeadlineSeconds and UntilTerminal clarifies the endpoint (terminal = phase is terminal) and uses language we are trying to emphasize.

This raises another question, whether activeDeadlineSeconds is expected to interact with this field - if both are specified, which definition of active is more correct (deadline for active applies to the node behavior, this new field applies to the controller, and there can be a significant delta between those two perspectives)?

Reply:
Somewhat.  

https://github.com/kubernetes/enhancements/pull/3931#discussion_r1156050560

Reply:
> This raises another question, whether activeDeadlineSeconds is expected to interact with this field - if both are specified, which definition of active is more correct (deadline for active applies to the node behavior, this new field applies to the controller, and there can be a significant delta between those two perspectives)?

So generally, if a user sets activeDeadliineSeconds, we would then go ahead and terminate the pod.  I'm unclear how it works if a pod gets killed by DeadlineExceeded will it start a new pod?  Or will we turn off `BackoffLimit` if a pod hits a deadline?

Regardless of that question, if `BackOffLimit` does retry a pod then we would start a new pod while the terminating pod is being terminated.  If a pod takes a while to terminate we could have another one running.  

I think that `activeDeadlineSeconds` does not include terminating in the deadline count.  If a Pod goes over the deadline than the job controller deletes those pods.

Reply:
adjacent thread regarding the name and the shape of the field: https://github.com/kubernetes/enhancements/pull/3940#discussion_r1161736497

Reply:
If you are adding a new field `terminating` in the status, then `activeUntilTerminal` is not accurate. The pods will be counted in a separate counter.

We could use a field name that explains when a pod is replaced:

```yaml
recreatePodsWhen: TerminatingOrFailed|Failed
```

Reply:
recreatePodsWhen: TerminatingOrFailed|Failed

How do you define an API with struct tags like this?

Reply:
I found out that this is short hand for a enum.

Reply:
updated to use `recreatePodsWhen` as it seems that Enum is the general view at this point.